snuba/admin/views.py

```diff
@@ -462,13 +520,53 @@ def clickhouse_trace_query() -> Response:
             "code": err.code,
         }
         return make_response(jsonify({"error": details}), 400)
+    except ClickHouseTimeoutError as err:
+        return make_response(
+            jsonify(
+                {
+                    "error": {
+                        "type": "clickhouse_timeout",
+                        "message": err.message or "Query timed out",
+                    }
+                }
+            ),
+            408,
+        )
     except Exception as err:
         return make_response(
             jsonify({"error": {"type": "unknown", "message": str(err)}}),
             500,
         )
 
 
+def parse_trace_for_query_ids(
+    trace_output: TraceOutput, storage_key: str
+) -> List[Dict[str, Any]]:
+    result = []
+    summarized_trace_output = trace_output.summarized_trace_output
+    storage_info = get_storage_info()
+    matched = next(
+        (info for info in storage_info if info["storage_name"] == storage_key), None
+    )
+    if matched is not None:
+        local_nodes = matched.get("local_nodes", [])
+        query_node = matched.get("query_node", None)
+        for node_name, query_summary in summarized_trace_output.query_summaries.items():
+            result.append(
+                {
+                    "host": local_nodes[0].get("host")
+                    if local_nodes
+                    else query_node.get("host"),  # type: ignore
+                    "port": local_nodes[0].get("port")
+                    if local_nodes
+                    else query_node.get("port"),  # type: ignore
+                    "query_id": query_summary.query_id,
+                    "node_name": node_name,
```

Comment:
I'm not so concerned about this in /admin/ code but since you know all the field names in advance, this feels more like a dataclass than a dictionary use case.

Reply:
Fair point. I converted this to a dataclass now.